source/uwp/Renderer/lib/ActionHelpers.cpp

```diff
@@ -0,0 +1,595 @@
+            THROW_IF_FAILED(renderContext->get_ElementRenderers(&elementRenderers));
+
+            ComPtr<IUIElement> buttonIcon;
+            ComPtr<IAdaptiveElementRenderer> elementRenderer;
+            THROW_IF_FAILED(elementRenderers->Get(HStringReference(L"Image").Get(), &elementRenderer));
+            if (elementRenderer != nullptr)
+            {
+                elementRenderer->Render(adaptiveCardElement.Get(), renderContext, childRenderArgs.Get(), &buttonIcon);
+                if (buttonIcon == nullptr)
+                {
+                    XamlHelpers::SetContent(localButton.Get(), title.Get());
+                    return;
+                }
+            }
+
+            // Create title text block
+            ComPtr<ITextBlock> buttonText =
+                XamlHelpers::CreateXamlClass<ITextBlock>(HStringReference(RuntimeClass_Windows_UI_Xaml_Controls_TextBlock));
+            THROW_IF_FAILED(buttonText->put_Text(title.Get()));
+            THROW_IF_FAILED(buttonText->put_TextAlignment(TextAlignment::TextAlignment_Center));
+
+            // Handle different arrangements inside button
+            ComPtr<IFrameworkElement> buttonIconAsFrameworkElement;
+            THROW_IF_FAILED(buttonIcon.As(&buttonIconAsFrameworkElement));
+            ComPtr<IUIElement> separator;
+            if (iconPlacement == ABI::AdaptiveNamespace::IconPlacement::AboveTitle && allActionsHaveIcons)
+            {
+                THROW_IF_FAILED(buttonContentsStackPanel->put_Orientation(Orientation::Orientation_Vertical));
+
+                // Set icon height to iconSize (aspect ratio is automatically maintained)
+                THROW_IF_FAILED(buttonIconAsFrameworkElement->put_Height(iconSize));
+            }
+            else
+            {
+                THROW_IF_FAILED(buttonContentsStackPanel->put_Orientation(Orientation::Orientation_Horizontal));
+
+                // Add event to the image to resize itself when the textblock is rendered
+                ComPtr<IImage> buttonIconAsImage;
+                THROW_IF_FAILED(buttonIcon.As(&buttonIconAsImage));
+
+                EventRegistrationToken eventToken;
+                THROW_IF_FAILED(buttonIconAsImage->add_ImageOpened(
+                    Callback<IRoutedEventHandler>([buttonIconAsFrameworkElement,
+                                                   buttonText](IInspectable* /*sender*/, IRoutedEventArgs * /*args*/) -> HRESULT {
+                        ComPtr<IFrameworkElement> buttonTextAsFrameworkElement;
+                        RETURN_IF_FAILED(buttonText.As(&buttonTextAsFrameworkElement));
+
+                        return SetMatchingHeight(buttonIconAsFrameworkElement.Get(), buttonTextAsFrameworkElement.Get());
+                    }).Get(),
+                    &eventToken));
+
+                // Only add spacing when the icon must be located at the left of the title
+                UINT spacingSize;
```

Comment:
>UINT [](start = 16, length = 4)

should we normalize to use UINT32?

Reply:
this was moved from its original location, so I don't want to mess with it here. that said, we should probably take a look at our `UINT`/`UINT32` usage

---
In reply to: [326325155](https://github.com/microsoft/AdaptiveCards/pull/3457#discussion_r326325155) [](ancestors = 326325155)